trinity/plugins/builtin/ethstats/plugin.py

```diff
@@ -56,12 +70,24 @@ def configure_parser(self, arg_parser: ArgumentParser, subparser: _SubParsersAct
         )
 
     def should_start(self) -> bool:
-        if not self.context.args.ethstats:
+        args = self.context.args
+
+        if not args.ethstats:
             return False
 
+        if (args.ethstats_server_url):
+            self.server_url = args.ethstats_server_url
+        else:
+            self.server_url = self.get_default_server_url()
+
+        self.server_secret = args.ethstats_server_secret
+
+        self.node_id = args.ethstats_node_id
+        self.node_contact = args.ethstats_node_contact
+
         configuration_provided: bool = all((
-            self.context.args.ethstats_server_url,
-            self.context.args.ethstats_server_secret,
+            self.server_url,
```

Comment:
Structurally this is sound, however, I'd like to request a refactor to the ordering of things.  Rather than doing all the setup and then checking whether all of the values are present, I'd prefer we invert this and do the checks first, trigger an early exit if there is a mis-configuration, and only set the values if everything is configured correctly.

For example, I believe the following are the error conditions:

- no `server_url` provided and `network_id` is not in known list.
- `--ethstats` was enabled but not `server_secret` was provided.

The modified function would look something like:

```python
if not args.ethstats:
    return

if not args.ethstats_server_url and self.context.trinity_config.network_id not in DEFAULT_SERVER_URLS:
    self.logger.error("You must provide the ethstats server url using the `--ethstats-server-url` option")
    self.context.shutdown_host()
    return
elif not args.ethstats_server_secret
    self.logger.error("You must provide secret for ethstats using `--ethstats-server-secret`")
    self.context.shutdown_host()
    return

# now we know we have everything we need and can set the various values on `self`
````

By pre-validating we avoid a certain family of errors which can easily show up in subsequent refactors.

Reply:
Do I need to call `shutdown_host` here? As ethstats plugin is not critical I think logging error that plugin won’t start should be enough.

Reply:
Refactored to be a bit more verbose with errors. Used `shutdown_host` as suggested.

Reply:
To answer the previous question, while you are correct that ethstats is not a critical plugin, it seems it would violate the "principle of least surprise" to allow the client to boot with one of these values missing.  It'd be really easy to miss the startup message after which the user would likely assume that ethstats was running smoothly.